### PR TITLE
Automatically check `DataModule.has_{setup,teardown,prepare_data}` [2/2]

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

This PR automates checking whether the appropriate setup/teardown/prepare_data has been called. The check was only performed by the `Trainer`. Now it is done by the `DataModule` itself.

This has implications for the users if they were manually calling these functions themselves, as they will now be a no-op if they had been called already.

```python
# user code:
dm.setup('fit')
...
dm.setup('fit')  # before this PR setup would get called it again, now it is a no-op
```

## Fixes:

Fixes the check when the stage is `None`. Before it wasn't checking each attribute (`has_setup_{fit,validate,test}`) but `has_setup_None` which is not defined.

This PR does NOT fix_ #7286 because to properly resolve it, we need to avoid `stage` becoming `None` in the first place.

## Note

The changes in this PR are the opposite of those discussed in https://github.com/PyTorchLightning/pytorch-lightning/issues/7301.

This PR attempts to fix the current design implementation. Not saying we should not change the design itself in the future.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: docs/source/extensions/datamodules.rst [hunk 1]
```diff
@@ -249,10 +245,17 @@ There are also data operations you might want to perform on every GPU. Use setup
...
+It is used to separate setup logic for ``trainer.{fit,validate,test}``. If ``setup`` is called with ``stage = None``,
+we assume all stages have been set-up.
 
+.. note:: ``setup`` is called from every process. Setting state here is okay.
 .. note:: ``teardown`` can be used to clean up the state. It is also called from every process
+.. note::
+    ``{setup,teardown,prepare_data}`` call will be only called once for a specific stage.
+    If the stage was ``None`` then we assume ``{fit,validate,test}`` have been called. For example, this means that
+    any duplicate ``dm.setup('fit')`` calls will be a no-op. To avoid this, you can overwrite
+    ``dm.has_setup_fit = False``
```
Comment (carmocca) [inline docs/source/extensions/datamodules.rst:258]:
```suggestion
    ``dm._has_setup_fit = False``
```

**Review by ananthsub — APPROVED** [review]
Agreed with the fix for now but I'm in favor of simplifying/deprecating these properties in v1.4

File: pytorch_lightning/core/datamodule.py [hunk 2]
```diff
@@ -366,15 +366,25 @@ def wrapped_fn(*args: str, **kwargs: Optional[str]) -> Any:
...
+                    has_run = getattr(obj, attr)
+                    setattr(obj, attr, True)
 
             elif name == "prepare_data":
+                has_run = obj._has_prepared_data
                 obj._has_prepared_data = True
 
-            return fn(*args, **kwargs)
+            else:
+                raise ValueError(name)
```
Comment (awaelchli) [inline pytorch_lightning/core/datamodule.py:384]: I believe this is unreachable. See the test I added in the last commit. We get AttributeError before that.

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM ! @ananthsub I think this is an important feature as preparing a dataset can be costly and shouldn't be done several times.

How do you propose to solve this problem if we remove those properties ?

File: pytorch_lightning/core/datamodule.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/core/datamodule.py:384]: I removed the `ValueError`, it'd only appear if you were to do `dm.sunflower("fit")` but that would not work as we manually wrap these functions so it's pointless to have.

**Pull request merged**